server/server.go

```diff
@@ -2,202 +2,86 @@ package server
-	if err != nil {
-		s.logger.Error("failed to get mime", zap.Error(err))
-	}
-
-	w.Header().Set("Content-Type", mime)
-	w.Header().Set("Cache-Control", "no-store")
-
-	_, err = w.Write(image)
-	if err != nil {
-		s.logger.Error("failed to write image", zap.Error(err))
-	}
-}
-
-func (s *audioHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
-
-	messageIDs, ok := r.URL.Query()["messageId"]
-	if !ok || len(messageIDs) == 0 {
-		s.logger.Error("no messageID")
-		return
-	}
-	messageID := messageIDs[0]
-	var audio []byte
-	err := s.db.QueryRow(`SELECT audio_payload FROM user_messages WHERE id = ?`, messageID).Scan(&audio)
-	if err != nil {
-		s.logger.Error("failed to find image", zap.Error(err))
-		return
-	}
-	if len(audio) == 0 {
-		s.logger.Error("empty audio")
-		return
-	}
-
-	w.Header().Set("Content-Type", "audio/aac")
-	w.Header().Set("Cache-Control", "no-store")
+const (
+	basePath       = "/messages"
+	identiconsPath = basePath + "/identicons"
+	imagesPath     = basePath + "/images"
+	audioPath      = basePath + "/audio"
+)
 
-	_, err = w.Write(audio)
-	if err != nil {
-		s.logger.Error("failed to write audio", zap.Error(err))
-	}
-}
+var (
+	defaultIP = net.IP{127, 0, 0, 1}
+)
 
 type Server struct {
-	Port   int
+	port   int
 	run    bool
 	server *http.Server
 	logger *zap.Logger
 	db     *sql.DB
 	cert   *tls.Certificate
+	netIP  net.IP
+}
+
+type Config struct {
+	Cert  *tls.Certificate
+	NetIP net.IP
+	Port  int
 }
 
-func NewServer(db *sql.DB, logger *zap.Logger) (*Server, error) {
-	err := generateTLSCert()
+// NewServer returns a *Server. If the config param is nil the default Server values are applied to the new Server
+// otherwise the config params are applied to the Server.
+func NewServer(db *sql.DB, logger *zap.Logger, config *Config) (*Server, error) {
+	s := &Server{db: db, logger: logger}
 
-	if err != nil {
-		return nil, err
+	if config == nil {
+		err := generateTLSCert()
+		if err != nil {
+			return nil, err
+		}
+
+		s.cert = globalCertificate
+		s.netIP = defaultIP
+		s.port = 0
+	} else {
+		s.cert = config.Cert
+		s.netIP = config.NetIP
+		s.port = config.Port
 	}
 
-	return &Server{db: db, logger: logger, cert: globalCertificate, Port: 0}, nil
+	return s, nil
 }
 
 func (s *Server) listenAndServe() {
-	cfg := &tls.Config{Certificates: []tls.Certificate{*s.cert}, ServerName: "localhost", MinVersion: tls.VersionTLS12}
+	cfg := &tls.Config{Certificates: []tls.Certificate{*s.cert}, ServerName: s.netIP.String(), MinVersion: tls.VersionTLS12}
 
 	// in case of restart, we should use the same port as the first start in order not to break existing links
-	addr := fmt.Sprintf("localhost:%d", s.Port)
+	addr := fmt.Sprintf("%s:%d", s.netIP, s.port)
 
 	listener, err := tls.Listen("tcp", addr, cfg)
 	if err != nil {
 		s.logger.Error("failed to start server, retrying", zap.Error(err))
-		s.Port = 0
+		s.port = 0 //TODO find out why the port is set to 0 here
 		err = s.Start()
```

Comment:
because this error happens when something is already listening on that port. Setting to 0 takes another random port. This happens on iOS when the app is put in background the port is not released for a long time but the socket is closed, so you would enter an infinite loop trying to retake the previous port.

Reply:
Thank you very much. That is helpful to know, I'll change my code to handle randomly selected port numbers rather than hardcoding a specific port.

Reply:
@bitgamma I've change the way `Server` works, we no longer use `port` instead we rely on having an active `net.Listener`. This method is explicit that the listener should always choose the port number, we don't need to reset the port number, we just reset the listener. What do you think?

Reply:
I am ooo (vacation) so I cannot review in details, but in general make sure that the transition from fg/bg and back on both Android and iOS is not broken by this, cause they both have quirks. Also, is there the risk that if you leave an image or a chat open and lock the screen, when you resume the images will all be broken since the port changed. The reason why I used a random port but then tried to keep it was to avoid broken links

Reply:
Ah sorry, I didn't realise you were on holiday.

> The reason why I used a random port but then tried to keep it was to avoid broken links

Yes, this is what I've done basically. I really like the idea of the `net.Listener` choosing the port number. So instead of storing the port number on the `Server`, I've replaced that with storing the listener itself, as the listener is always the source of the port number.

Reply:
What I'll do is write some tests to check fg/bg transitioning doesn't produce any errors. My standard e2e test works fine, so that's a good staring point.

Reply:
ok, but e2e is not going to be enough, especially on iOS sockets are silently killed in some cases. This really needs to be tested on hw